### PR TITLE
Add DollarBraceEmbeddedVariableMigration

### PR DESCRIPTION
**Pull request by viratyosin**
Summary: This migration converts variables in string interpolation using the syntax `"${var}"` to use the syntax `"{$var}"`.

Test plan:
```
vendor/bin/hacktest tests/DollarBraceEmbeddedVariableMigrationTest.hack
```

### CODE REVIEW

**Comment by viratyosin** [comment]
One case that's not handled is trivial in the braces, e.g.
`"${ /* leading */ x /* or trailing */ }"` will result in the same thing as `"${x}"`.

Not sure if anyone does this, but if you think it's worth it, I can try to handle this case by preserving any trivia as trailing trivia (`"{$x}"` won't parse right if there's trivia between the left brace and the dollar).

File: src/Migrations/DollarBraceEmbeddedVariableMigration.hack [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+namespace Facebook\HHAST\Migrations;
+
+use namespace Facebook\HHAST;
+
+final class DollarBraceEmbeddedVariableMigration extends StepBasedMigration {
+  private static function migrateLiteralExpression(
+    HHAST\LiteralExpression $node,
+  ): HHAST\LiteralExpression {
+    $children = $node->getChildren();
+    $child = \HH\Lib\C\first($children);
```
Comment (fredemmott) [inline src/Migrations/DollarBraceEmbeddedVariableMigration.hack]: use namespace

File: src/Migrations/DollarBraceEmbeddedVariableMigration.hack [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    if (!($child is HHAST\NodeList<_>)) {
```
Comment (fredemmott) [inline src/Migrations/DollarBraceEmbeddedVariableMigration.hack]: until D15890968 is landed, this should be an instanceof

File: src/Migrations/DollarBraceEmbeddedVariableMigration.hack [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      return $node;
+    }
+    $literal_parts = $child->getChildren();
+    $first_part = \HH\Lib\C\first($literal_parts);
+    if (
+      !(
+        $first_part is HHAST\DoubleQuotedStringLiteralHeadToken ||
```
Comment (fredemmott) [inline src/Migrations/DollarBraceEmbeddedVariableMigration.hack]: IStringLiteral

File: src/Migrations/DollarBraceEmbeddedVariableMigration.hack [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    } else {
+      return $node;
+    }
+  }
+
+  <<__Override>>
+  final public function getSteps(): Traversable<IMigrationStep> {
+    return vec[
+      new TypedMigrationStep(
+        'covert "${foo}" to "{$foo}"',
```
Comment (fredemmott) [inline src/Migrations/DollarBraceEmbeddedVariableMigration.hack]: convert

File: src/__Private/MigrationCLI.hack [hunk 5]
```diff
@@ -78,6 +79,13 @@ class MigrationCLI extends CLIWithRequiredArguments {
         'Migrate nullable shape fields to be both nullable and optional',
         '--optional-shape-fields',
       ),
+      CLIOptions\flag(
+        () ==> {
+          $this->migrations[] = DollarBraceEmbeddedVariableMigration::class;
+        },
+        'Migreate instances of "${foo}" to "{$foo}"',
```
Comment (fredemmott) [inline src/__Private/MigrationCLI.hack]: Migrate

File: src/__Private/MigrationCLI.hack [hunk 6]
```diff
@@ -78,6 +79,13 @@ class MigrationCLI extends CLIWithRequiredArguments {
...
+        '--dollar-brace-variable',
```
Comment (fredemmott) [inline src/__Private/MigrationCLI.hack]: dollar-brace-variable-interpolation maybe? #bikeshed

File: tests/examples/migrations/dollar_brace_embedded_variable.php.in [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<?hh // strict
+
+/**
+ * Copyright (c) 2016, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional
+ * grant of patent rights can be found in the PATENTS file in the same
+ * directory.
```
Comment (fredemmott) [inline tests/examples/migrations/dollar_brace_embedded_variable.php.in]: Use the same MIT header as the other files in this diff.

HHAST shouldn't have any BSD+PATENTS any more - but I see there's a few, I'll fix up the current ones separately.

File: src/Migrations/DollarBraceEmbeddedVariableMigration.hack (hunk 3 above)
Comment (viratyosin) [reply src/Migrations/DollarBraceEmbeddedVariableMigration.hack]: I meant to exclude the other node types that implement `IStringLiteral` e.g. `SingleQuotedStringLiteralToken`

**Comment by fredemmott** [comment]
This needs a rebase on top of e9fb7f6

**Pull request merged**